src/host/readDataCooked.cpp

```diff
@@ -436,633 +196,1027 @@ bool COOKED_READ_DATA::AtEol() const noexcept
-            }
+    if (wch >= L'0' && wch <= L'9')
+    {
+        if (popup.commandNumber.bufferSize < CommandNumberMaxInputLength)
+        {
+            popup.commandNumber.buffer[popup.commandNumber.bufferSize++] = wch;
+        }
+    }
+    else if (wch == UNICODE_BACKSPACE)
+    {
+        if (popup.commandNumber.bufferSize > 0)
+        {
+            popup.commandNumber.buffer[--popup.commandNumber.bufferSize] = L' ';
         }
     }
+    else
+    {
+        return;
+    }
 
-    const auto inputSizeBefore = input.size();
-    GetInputBuffer()->Consume(isUnicode, input, writer);
+    RowWriteState state{
+        .text = { popup.commandNumber.buffer.data(), CommandNumberMaxInputLength },
+        .columnBegin = popup.contentRect.right - CommandNumberMaxInputLength,
+        .columnLimit = popup.contentRect.right,
+    };
+    _screenInfo.GetTextBuffer().Write(popup.contentRect.top, _screenInfo.GetPopupAttributes(), state);
+}
+
+void COOKED_READ_DATA::_popupHandleCommandListInput(Popup& popup, const wchar_t wch, const uint16_t vkey, const DWORD modifiers)
+{
+    auto& cl = popup.commandList;
 
-    if (LineCount > 1)
+    if (wch == UNICODE_CARRIAGERETURN)
     {
-        // This is a continuation of the above identical if condition.
-        // We've truncated the `input` slice and now we need to restore it.
-        const auto inputSizeAfter = input.size();
-        const auto amountConsumed = inputSizeBefore - inputSizeAfter;
-        input = { _backupLimit, _bytesRead / sizeof(wchar_t) };
-        input = input.substr(std::min(input.size(), amountConsumed));
-        GetInputReadHandleData()->SaveMultilinePendingInput(input);
+        _buffer.assign(_history->RetrieveNth(cl.selected));
+        _popupDone();
+        return _handleChar(UNICODE_CARRIAGERETURN, modifiers);
     }
-    else if (!input.empty())
+
+    switch (vkey)
     {
-        GetInputReadHandleData()->SavePendingInput(input);
+    case VK_ESCAPE:
+        _popupDone();
+        return false;
+    case VK_F9:
+        _popupPush(PopupKind::CommandNumber);
+        return false;
+    case VK_DELETE:
+        _history->Remove(cl.selected);
+        if (_history->GetNumberOfCommands() <= 0)
+        {
+            _popupDone();
+            return false;
+        }
+        break;
+    case VK_LEFT:
+    case VK_RIGHT:
+        _replaceBuffer(_history->RetrieveNth(cl.selected));
+        _popupDone();
+        return false;
+    case VK_UP:
+        if (WI_IsFlagSet(modifiers, SHIFT_PRESSED))
+        {
+            _history->Swap(cl.selected, cl.selected - 1);
+        }
+        cl.selected--;
+        break;
+    case VK_DOWN:
+        if (WI_IsFlagSet(modifiers, SHIFT_PRESSED))
+        {
+            _history->Swap(cl.selected, cl.selected + 1);
+        }
+        cl.selected++;
+        break;
+    case VK_HOME:
+        cl.selected = 0;
+        break;
+    case VK_END:
+        cl.selected = INT_MAX;

```

Comment:
maybe leave the note about 

```
// Move waaay forward, _popupDrawCommandList() can handle it
```

Reply:
cause that at least clues you in that _someone else_ is about to clamp this stuff